linker/shared/src/main/scala/org/scalajs/linker/frontend/Refiner.scala

```diff
@@ -133,22 +136,31 @@ private object Refiner {
     }
   }
 
-  private class ClassInfoCache {
+  private class ClassInfoCache(checkIR: Boolean) {
     private var cacheUsed: Boolean = false
     private val methodsInfoCaches = MethodDefsInfosCache()
     private val jsConstructorInfoCache = new JSConstructorDefInfoCache()
     private val exportedMembersInfoCaches = JSMethodPropDefsInfosCache()
     private var info: Infos.ClassInfo = _
 
-    def loadInfo(classDef: ClassDef)(implicit ec: ExecutionContext): Future[Infos.ClassInfo] = Future {
-      update(classDef)
+    def loadInfo(classDef: ClassDef, logger: Logger)(implicit ec: ExecutionContext): Future[Infos.ClassInfo] = Future {
+      update(classDef, logger)
       info
     }
 
-    private def update(classDef: ClassDef): Unit = synchronized {
+    private def update(classDef: ClassDef, logger: Logger): Unit = synchronized {
       if (!cacheUsed) {
         cacheUsed = true
 
+        if (checkIR) {
+          val errorCount = ClassDefChecker.check(classDef,
+              allowReflectiveProxies = true, allowTransients = true, logger)
+          if (errorCount != 0) {
+            throw new AssertionError(s"There were $errorCount ClassDef " +
+                "checking errors after optimizing. Please report this as a bug.")
```

Comment:
To keep sentences on a single line (better grepping):
```suggestion
            throw new AssertionError(
                s"There were $errorCount ClassDef checking errors after optimizing. " +
                "Please report this as a bug.")
```